spinalcordtoolbox/image.py

```diff
@@ -279,7 +282,22 @@ def __init__(self, param=None, hdr=None, orientation=None, absolutepath=None, di
 
         # load an image from file
         if isinstance(param, str):
-            self.loadFromPath(param, verbose)
+            try:
+                self.loadFromPath(param, mmap, verbose)
+            except OSError as e:
+                if str(e) == "[Errno 24] Too many open files":
+                    try:
+                        # Only load images into memory if there are no more available file descriptors
+                        self.loadFromPath(param, False, verbose)
+                    except OSError as e:
+                        # If for some reason, an OSError is still thrown, then give the user a helpful error message
+                        if str(e) == "[Errno 24] Too many open files":
+                            raise OSError("[Errno 24] Too many open files. Please try increasing your system's file "
+                                          "descriptor limit by using the command `ulimit -Sn`.")
+                        else:
+                            raise e
```

Comment:
One thing I want to note here is, if we are catching exceptions and replacing the tracebacks with a friendlier output message, then presumably that would be the job of a non-exception function like `printv`.

But, we've also got a soft-ban on `printv` in the API. So, I guess I should be using `logger.error` + `sys.exit` instead?

Reply:
I think raising an exception is appropriate here. The `Image` constructor lives in the API layer, so it doesn't have the whole picture, and `sys.exit` would be a bit rude.

If we want to try a non-mmap approach when encountering the "Too many open files" error, I think this would normally be handled in the calling code, which actually has access to the other open files and can close them before retrying. This is only possible by throwing an exception, which can be caught by the calling code. And if it's not caught, at least there'll be a useful suggestion displayed to the user in the traceback.